## Introduction to the Data

Hacker News is a community where users can submit articles, and other users can upvote those articles. The articles with the most upvotes make it to the front page, where they're more visible to the community.

The data set consists of submissions users made to Hacker News from 2006 to 2015.

The columns in the dataset include:

- submission_time - When the article was submitted
- upvotes - The number of upvotes the article received
- url - The base URL of the article
- headline - The article's headline

In this project, I'll be predicting the number of upvotes the articles received, based on their headlines. 

Because upvotes are an indicator of popularity, we'll discover which types of articles tend to be the most popular.

In [19]:
import pandas as pd
submissions = pd.read_csv("sel_hn_stories.csv")
submissions.columns = ["submission_time", "upvotes", "url", "headline"]

In [20]:
submissions

submission_time  upvotes                            url  \
0         2010-02-17T16:57:59Z        1            blog.jonasbandi.net   
1         2014-02-04T02:36:30Z        1                  blogs.wsj.com   
2         2011-10-26T07:11:29Z        1                 threatpost.com   
3         2011-04-03T15:43:44Z       67               algorithm.com.au   
4         2013-01-13T16:49:20Z        1                winmacsofts.com   
5         2013-09-04T12:10:52Z        1     theincidentaleconomist.com   
6         2012-03-09T20:25:42Z        1              worldometers.info   
7         2010-04-22T13:23:10Z       26                       docs.com   
8         2012-05-06T16:08:46Z        2             blog.hackplanet.in   
9     2014-12-23T00:55:31.000Z        1                  curt-rice.com   
10    2014-08-14T14:27:07.000Z       26                            NaN   
11        2013-06-27T20:35:34Z        8                    nytimes.com   
12        2014-03-23T00:22:26Z        1                 retargeter.com   
13        2011-12-14T15:50:25Z        2                extremetech.com   
14        2011-10-01T05:28:30Z        9  tenprinciplesofgooddesign.com   
15        2012-04-17T17:02:58Z      109                     google.com   
16        2012-04-02T18:14:03Z       48                    tomayko.com   
17    2014-07-28T06:53:09.000Z        3             washingtonpost.com   
18        2010-08-18T02:47:01Z        3                    youtube.com   
19        2011-12-02T15:12:30Z        1           changerlaplanete.com   
20        2011-07-05T10:32:02Z       61         holdenweb.blogspot.com   
21        2013-02-09T23:18:04Z        1             peninsulapress.com   
22        2012-06-04T21:43:00Z        1                  letitlinks.ru   
23        2012-03-12T00:55:44Z       10                   usatoday.com   
24        2011-11-10T17:32:14Z       15                    nytimes.com   
25        2010-01-12T14:55:24Z        1                 rowtheboat.com   
26        2011-03-08T02:54:22Z        1        join.latinosexdates.com   
27        2012-04-03T23:12:51Z        1               androidandme.com   
28        2014-04-14T13:37:13Z        1                     kottke.org   
29        2010-12-12T12:48:48Z        2                    nytimes.com   
...                        ...      ...                            ...   
2969      2010-02-16T16:40:05Z        5               smexperiment.com   
2970      2011-10-11T11:47:02Z        2                    youtube.com   
2971      2013-08-13T21:18:00Z        1                  smitpatel.com   
2972  2015-01-12T14:12:46.000Z        3                 chronicons.com   
2973      2013-04-04T14:09:41Z        1                        veri.fi   
2974      2013-11-13T19:03:32Z        1                raspberrypi.org   
2975      2010-07-12T13:00:36Z        2                       gist.com   
2976      2013-02-23T04:39:10Z        1               dunkhighclub.com   
2977      2011-12-20T00:04:49Z        1                        npr.org   
2978      2013-05-10T13:57:01Z        1           devops.rackspace.com   
2979  2014-11-06T10:49:34.000Z        1               ariannaodell.com   
2980      2011-12-18T16:24:33Z        9                 techcrunch.com   
2981      2010-10-29T22:19:52Z       32                 techcrunch.com   
2982      2009-01-14T17:27:28Z        1                    nettuts.com   
2983  2014-06-30T09:10:36.000Z        2                    9to5mac.com   
2984      2011-11-12T06:54:56Z        1       thomassabo-charmshop.com   
2985      2011-10-07T19:23:13Z      172                       b3ta.com   
2986      2014-02-25T01:40:44Z        2              blog.ethereum.org   
2987      2012-07-29T12:44:54Z        7                      bbc.co.uk   
2988      2008-03-11T01:44:45Z        7                     k.blog.com   
2989      2011-12-15T21:56:28Z        2        blog.millermedeiros.com   
2990      2012-04-26T10:55:04Z      170                jasondavies.com   
2991      2011-12-0

## Finding missing Values

In [21]:
submissions.isnull().sum()

submission_time      0
upvotes              0
url                189
headline            10
dtype: int64

In [22]:
submissions.notnull().sum()

submission_time    2999
upvotes            2999
url                2810
headline           2989
dtype: int64

Since url and headline NaN values constitutes about less than 7% of the entire datapoints, its fair to drop all rows that contains any NaN value

In [23]:
submissions = submissions.dropna(axis=0)

In [24]:
submissions.isnull().sum()

submission_time    0
upvotes            0
url                0
headline           0
dtype: int64

## Tokenizing & Preprocessing the Headlines

As earlier said, I'll be using only the Headlines column to train the model

In [25]:
tokenized_headlines = []
for item in submissions["headline"]:
    tokenized_headlines.append(item.split())
tokenized_headlines

[['Software:',
  'Sadly',
  'we',
  'did',
  'adopt',
  'from',
  'the',
  'construction',
  'analogy'],
 ['Google’s',
  'Stock',
  'Split',
  'Means',
  'More',
  'Control',
  'for',
  'Larry',
  'and',
  'Sergey'],
 ['SSL',
  'DOS',
  'attack',
  'tool',
  'released',
  'exploiting',
  'negotiation',
  'overhead'],
 ['Immutability', 'and', 'Blocks', 'Lambdas', 'and', 'Closures'],
 ['Comment', 'optimiser', 'la', 'vitesse', 'de', 'Wordpress?'],
 ['ilk', 'is', 'not', 'as', 'good', 'for', 'you', 'as', 'you', 'think'],
 ['Worldometers', '-', 'Real', 'time', 'world', 'statistics'],
 ['icrosoft', 'strikes', 'back:', 'introduces', 'docs', 'for', 'facebook'],
 ['Net', 'HTTP', 'status', 'codes'],
 ['Anecdata',
  'or',
  'how',
  'McKinsey’s',
  'story',
  'became',
  'Sheryl',
  'Sandberg’s',
  'fact'],
 ['Immigration', 'Overhaul', 'Passes', 'in', 'Senate'],
 ['What', 'matters', 'most', 'at', 'Ad:TECH', 'SF', '2014'],
 ['Amazon',
  'Silk',
  'revisited:',
  'Is',
  'the',
  'split',
  'cloud',

from tokenized_headlines:

- some punctuations are attached to each splitted item
- the same words are differentiated by uppercase and lowercase

In [26]:
punctuation = [",", ":", ";", ".", "'", '"', "’", "?", "/", "-", "+", "&", "(", ")"]
clean_tokenized = []
for item in tokenized_headlines:
    tokens = []
    for token in item:
        token = token.lower()
        for punc in punctuation:
            token = token.replace(punc, "")
        tokens.append(token)
    clean_tokenized.append(tokens)

In [27]:
clean_tokenized

[['software',
  'sadly',
  'we',
  'did',
  'adopt',
  'from',
  'the',
  'construction',
  'analogy'],
 ['googles',
  'stock',
  'split',
  'means',
  'more',
  'control',
  'for',
  'larry',
  'and',
  'sergey'],
 ['ssl',
  'dos',
  'attack',
  'tool',
  'released',
  'exploiting',
  'negotiation',
  'overhead'],
 ['immutability', 'and', 'blocks', 'lambdas', 'and', 'closures'],
 ['comment', 'optimiser', 'la', 'vitesse', 'de', 'wordpress'],
 ['ilk', 'is', 'not', 'as', 'good', 'for', 'you', 'as', 'you', 'think'],
 ['worldometers', '', 'real', 'time', 'world', 'statistics'],
 ['icrosoft', 'strikes', 'back', 'introduces', 'docs', 'for', 'facebook'],
 ['net', 'http', 'status', 'codes'],
 ['anecdata',
  'or',
  'how',
  'mckinseys',
  'story',
  'became',
  'sheryl',
  'sandbergs',
  'fact'],
 ['immigration', 'overhaul', 'passes', 'in', 'senate'],
 ['what', 'matters', 'most', 'at', 'adtech', 'sf', '2014'],
 ['amazon',
  'silk',
  'revisited',
  'is',
  'the',
  'split',
  'cloud',
  'brows

#### Assembling a Matrix of Unique Words

In [29]:
import numpy as np
unique_tokens = []
single_tokens = []
for tokens in clean_tokenized:
    for token in tokens:
        if token not in single_tokens:
            single_tokens.append(token)
        elif token in single_tokens and token not in unique_tokens:
            unique_tokens.append(token)

counts = pd.DataFrame(0, index=np.arange(len(clean_tokenized)), columns=unique_tokens)

In [31]:
for i, item in enumerate(clean_tokenized):
    for token in item:
        if token in unique_tokens:
            counts.iloc[i][token] += 1

#### Removing stopwords 

In [32]:
word_counts = counts.sum(axis=0)

counts = counts.loc[:,(word_counts >= 5) & (word_counts <= 100)]

## Training a Model with LInearRegression

first of all, I'll split the data into train and test data and out target column is the upvotes column

In [34]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(counts, submissions["upvotes"], test_size=0.35, random_state=1)

In [35]:
from sklearn.linear_model import LinearRegression

clf = LinearRegression()
clf.fit(X_train, y_train)

predictions = clf.predict(X_test)

#### Picking an Error Metric

I'll use rmse as thee error metric

In [41]:
mse = sum((predictions - y_test) ** 2) / len(predictions)
rmse = np.sqrt(mse)
rmse

54.556993831940254

## Training a Model with RandomForestRegressor

In [49]:
from sklearn.ensemble import RandomForestRegressor


In [47]:
rfr = RandomForestRegressor(n_estimators=5, random_state=1, min_samples_leaf=2)

rfr.fit(X_train, y_train)

predictions_1 = rfr.predict(X_test)

In [48]:
mse = sum((predictions_1 - y_test) ** 2) / len(predictions_1)
rmse = np.sqrt(mse)
rmse

44.84857168288281

 There's no hard and fast rule about what a "good" error rate is, because it depends on the problem we're solving and our error tolerance.

In this case, the mean number of upvotes is 10, and the standard deviation is 39.5.

Using the Linear Regression Model, we got our rmse to be 54.5. This means that our average error is 54.5 upvotes away from the true value. 

Using the RandomForest Regressor Model, we got our rmse to be 44.8. This means that our average error is 44.8 upvotes away from the true value


it seems the RandomForest Regressor Model performed better than Linear Regression Model, cause it's less farther from the standard deviation